### ISSUE TITLE
Union index signature parameter type has unclear / incorrect error 

### ISSUE THREAD

**Issue by catb0t** [microsoft/TypeScript#33427]
**TypeScript Version:**  3.7.0-dev.20190913

**Search Terms:**

* index signature parameter type 
* index signature union 

**Code**

```ts
interface A {
  [id : string | number] : boolean;
};
```

**Expected behavior:**

Either the following error message is displayed, or the error message makes it clear that **only one of** `string` xor `number` is permissible as an index signature parameter type. 

https://github.com/microsoft/TypeScript/blob/038d95144d8b93c2799d1732181c89c3d84362d5/src/compiler/diagnosticMessages.json#L962-L965

**Actual behavior:**

```
error TS1023: An index signature parameter type must be 'string' or 'number'.
```
To me, the expression `string | number` is equivalent to the phrase "'string' or 'number'", but the phrasing is not precise.

Either the wrong error is being thrown, or this TS1023 should be clarified to say "...must be **one of** 'string' or 'number'." or "...must be **either** 'string' or 'number'."

https://github.com/microsoft/TypeScript/blob/038d95144d8b93c2799d1732181c89c3d84362d5/src/compiler/diagnosticMessages.json#L74-L77

**Playground Link:** 
(error not shown in Playground)

**Related Issues:** 

https://github.com/microsoft/TypeScript/issues/2611
https://github.com/microsoft/TypeScript/issues/2049
https://github.com/microsoft/TypeScript/issues/8711 
https://github.com/microsoft/TypeScript/issues/18759

I will fix this in a PR if you tell me whether TS1023 is the right error with a fixed message, or if TS1337 should be thrown instead of TS1023.

**Comment by DanielRosenwasser** [comment]
I think you are correct that we should give that error message. Feel free to send a PR!

**Comment by fatcerberus** [comment]
> An index signature parameter type cannot be a union type. Consider using a mapped object type instead.

I personally really like this one, as it's been my experience that this is generally what people are trying to do when they write such a signature.  Alternatively, it could suggest using `Record`.

**Comment by DanielRosenwasser** [comment]
The current change is a good improvement and we can iterate based on feedback. Thanks @TroyTae!
